src/java/main/org/apache/zookeeper/server/quorum/ObserverMaster.java

```diff
@@ -0,0 +1,513 @@
+        return self.tick.get() + self.syncLimit;
+    }
+
+    @Override
+    public int getTickOfInitialAckDeadline() {
+        return self.tick.get() + self.initLimit + self.syncLimit;
+    }
+
+    @Override
+    public long getAndDecrementFollowerCounter() {
+        return followerCounter.getAndDecrement();
+    }
+
+    @Override
+    public void waitForEpochAck(long sid, StateSummary ss) throws IOException, InterruptedException {
+        // since this is done by an active follower, we don't need to wait for anything
+    }
+
+    @Override
+    public LearnerSnapshotThrottler getLearnerSnapshotThrottler() {
+        return learnerSnapshotThrottler;
+    }
+
+    @Override
+    public void waitForStartup() throws InterruptedException {
+        // since this is done by an active follower, we don't need to wait for anything
+    }
+
+    @Override
+    synchronized public long getLastProposed() {
+        return lastProposedZxid;
+    }
+
+    @Override
+    public long getEpochToPropose(long sid, long lastAcceptedEpoch) throws InterruptedException, IOException {
+        return self.getCurrentEpoch();
+    }
+
+    @Override
+    public ZKDatabase getZKDatabase() {
+        return zks.getZKDatabase();
+    }
+
+    @Override
+    public void waitForNewLeaderAck(long sid, long zxid) throws InterruptedException {
+        // no need to wait since we are a follower
+    }
+
+    @Override
+    public int getCurrentTick() {
+        return self.tick.get();
+    }
+
+    @Override
+    public void processAck(long sid, long zxid, SocketAddress localSocketAddress) {
+        if ((zxid & 0xffffffffL) == 0) {
+            /*
+             * We no longer process NEWLEADER ack by this method. However,
+             * the learner sends ack back to the leader after it gets UPTODATE
+             * so we just ignore the message.
+             */
+            return;
+        }
+
+        throw new RuntimeException("Observers shouldn't send ACKS ack = " + Long.toHexString(zxid));
+    }
+
+    @Override
+    public void touch(long sess, int to) {
+        zks.getSessionTracker().touchSession(sess, to);
+    }
+
+    boolean revalidateLearnerSession(QuorumPacket qp) throws IOException {
+        ByteArrayInputStream bis = new ByteArrayInputStream(qp.getData());
+        DataInputStream dis = new DataInputStream(bis);
+        long id = dis.readLong();
+        boolean valid = dis.readBoolean();
+        Iterator<Revalidation> itr = pendingRevalidations.iterator();
+        if (!itr.hasNext()) {
+            // not a learner session, handle locally
+            return false;
+        }
+        Revalidation revalidation = itr.next();
+        if (revalidation.sessionId != id) {
+            // not a learner session, handle locally
+            return false;
+        }
+        itr.remove();
+        LearnerHandler learnerHandler = revalidation.handler;
+        QuorumPacket deepCopy = new QuorumPacket(qp.getType(), qp.getZxid(),
```

Comment:
Brian, can you add a comment here to mention that we need a deepCopy of the QuorumPacket because qp object is reused to reduce the GC impact in learner when receiving packet from leader, and if we use that packet when it's being sent in the LearnerHandler asynchronously this packet might already be changed.